File: dialer/internal/audit/writer.go
package audit

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"
)

// Table identifies one of the five immutable audit tables.
type Table string

const (
	TableAuditLog        Table = "audit_log"
	TableCallWindowAudit Table = "call_window_audit"
	TableOriginateAudit  Table = "originate_audit"
	TableConsentLog      Table = "consent_log"
	TableDncSyncLog      Table = "dnc_sync_log"
)

// Result is returned by Append after a successful INSERT.
type Result struct {
	// ID is the auto-increment id of the newly inserted row.
	ID int64
	// RowHash is the 64-char lowercase hex SHA-256 row_hash computed by the
	// BEFORE INSERT trigger and read back via LAST_INSERT_ID() + SELECT.
	RowHash string
}

// Writer inserts rows into the immutable audit tables.
// The BEFORE INSERT trigger in MySQL handles prev_hash / row_hash / hash_at.
//
// Phase 1: direct MySQL INSERT via *sql.DB.
// Phase 4: batched via Valkey stream (no API change to callers).
type Writer struct {
	db *sql.DB
}

// NewWriter creates a new Writer backed by the given *sql.DB.
func NewWriter(db *sql.DB) *Writer {
	return &Writer{db: db}
}

// AppendAuditLog inserts a row into audit_log and returns its id + row_hash.
// The caller's transaction (if any) must be passed via ctx — Phase 1 uses
// auto-committed inserts for simplicity; Phase 4 may accept a *sql.Tx.
func (w *Writer) AppendAuditLog(ctx context.Context, r AuditLogRow) (Result, error) {
	var beforeJSON, afterJSON *string
	if r.BeforeJSON != nil {
		b, _ := json.Marshal(r.BeforeJSON)
		s := string(b)
		beforeJSON = &s
	}
	if r.AfterJSON != nil {
		b, _ := json.Marshal(r.AfterJSON)
		s := string(b)
		afterJSON = &s
	}

	result, err := w.db.ExecContext(ctx, `
		INSERT INTO audit_log
		  (tenant_id, actor_user_id, actor_kind, action, entity_type, entity_id,
		   before_json, after_json, request_id, ip_address, user_agent, ts)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		r.TenantID, optUint64(r.ActorUserID), r.ActorKind,
		r.Action, r.EntityType, r.EntityID,
		beforeJSON, afterJSON,
		r.RequestID, r.IPAddress, r.UserAgent,
		r.Ts.UTC().Format("2006-01-02 15:04:05.000000"),
	)
	if err != nil {
		return Result{}, fmt.Errorf("audit.Writer.AppendAuditLog: %w", err)
	}

	id, err := result.LastInsertId()
	if err != nil {
		return Result{}, fmt.Errorf("audit.Writer.AppendAuditLog: last insert id: %w", err)
	}

	rowHash, err := w.readRowHash(ctx, "audit_log", id)
	if err != nil {
		return Result{}, err
	}
	return Result{ID: id, RowHash: rowHash}, nil
}

// AppendCallWindowAudit inserts a row into call_window_audit.
func (w *Writer) AppendCallWindowAudit(ctx context.Context, r CallWindowAuditRow) (Result, error) {
	result, err := w.db.ExecContext(ctx, `
		INSERT INTO call_window_audit
		  (tenant_id, lead_id, phone_e164, campaign_id, decision, reason,
		   tz_iana, tz_confidence, state_code, zip, party_local, party_dow,
		   effective_open_min, effective_close_min, rule_applied,
		   enforcement_point, next_open_at, call_uuid)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		r.TenantID, r.LeadID, r.PhoneE164, r.CampaignID,
		r.Decision, r.Reason,
		r.TzIana, r.TzConfidence, r.StateCode, r.Zip,
		optTime(r.PartyLocal), optInt(r.PartyDow),
		optInt(r.EffectiveOpenMin), optInt(r.EffectiveCloseMin),
		r.RuleApplied, r.EnforcementPoint, optTime(r.NextOpenAt), r.CallUUID,
	)
	if err != nil {
		return Result{}, fmt.Errorf("audit.Writer.AppendCallWindowAudit: %w", err)
	}
	id, _ := result.LastInsertId()
	rowHash, err := w.readRowHash(ctx, "call_window_audit", id)
	if err != nil {
		return Result{}, err
	}
	return Result{ID: id, RowHash: rowHash}, nil
}

// AppendConsentLog inserts a row into consent_log.
func (w *Writer) AppendConsentLog(ctx context.Context, r ConsentLogRow) (Result, error) {
	result, err := w.db.ExecContext(ctx, `
		INSERT INTO consent_log
		  (tenant_id, call_uuid, lead_id, phone_e164, prompt_id,
		   dtmf_response, outcome, language, prompt_played_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		r.TenantID, r.CallUUID, r.LeadID, r.PhoneE164, r.PromptID,
		r.DtmfResponse, r.Outcome, r.Language,
		r.PromptPlayedAt.UTC().Format("2006-01-02 15:04:05.000000"),
	)
	if err != nil {
		return Result{}, fmt.Errorf("audit.Writer.AppendConsentLog: %w", err)
	}
	id, _ := result.LastInsertId()
	rowHash, err := w.readRowHash(ctx, "consent_log", id)
	if err != nil {
		return Result{}, err
	}
	return Result{ID: id, RowHash: rowHash}, nil
}

// AppendDncSyncLog inserts a row into dnc_sync_log.
func (w *Writer) AppendDncSyncLog(ctx context.Context, r DncSyncLogRow) (Result, error) {
	result, err := w.db.ExecContext(ctx, `
		INSERT INTO dnc_sync_log
		  (source, kind, file_hash, added, removed, started_at, completed_at)
		VALUES (?, ?, ?, ?, ?, ?, ?)`,
		r.Source, r.Kind, r.FileHash,
		r.Added, r.Removed,
		r.StartedAt.UTC().Format("2006-01-02 15:04:05.000000"),
		optTimeStr(r.CompletedAt),
	)
	if err != nil {
		return Result{}, fmt.Errorf("audit.Writer.AppendDncSyncLog: %w", err)
	}
	id, _ := result.LastInsertId()
	rowHash, err := w.readRowHash(ctx, "dnc_sync_log", id)
	if err != nil {
		return Result{}, err
	}
	return Result{ID: id, RowHash: rowHash}, nil
}

// readRowHash fetches the row_hash computed by the BEFORE INSERT trigger.
func (w *Writer) readRowHash(ctx context.Context, table Table, id int64) (string, error) {
	var rowHash string
	err := w.db.QueryRowContext(ctx,
		fmt.Sprintf("SELECT row_hash FROM `%s` WHERE id = ?", table), id,
	).Scan(&rowHash)
	if err != nil {
		return "", fmt.Errorf("audit.Writer.readRowHash(%s, %d): %w", table, id, err)
	}
	return rowHash, nil
}

// ---------------------------------------------------------------------------
// Nullable helpers
// ---------------------------------------------------------------------------

func optUint64(v *uint64) interface{} {
	if v == nil {
		return nil
	}
	return *v
}

func optInt(v *int) interface{} {
	if v == nil {
		return nil
	}
	return *v
}

func optTime(v *time.Time) interface{} {
	if v == nil {
		return nil
	}
	return v.UTC().Format("2006-01-02 15:04:05.000000")
}

func optTimeStr(v *time.Time) interface{} {
	return optTime(v)
}

File: api/src/services/audit/writer.ts
/**
 * api/src/services/audit/writer.ts — AuditWriter
 *
 * Low-latency, chain-preserving append to any immutable audit table.
 *
 * Phase 1 contract (PLAN §10.2):
 *   - Validates input with per-table Zod schema (no 0x1F, no NUL, ≤4 KB JSON)
 *   - Issues the INSERT via Prisma (trigger fires, hash chain extended)
 *   - Returns { id, rowHash } to caller
 *   - On error: bubbles up — caller MUST propagate (no silent drop)
 *
 * The INSERT path uses $queryRaw for tables not natively modelled in Prisma
 * because the chain columns are set by the DB trigger (not by Prisma logic).
 * For audit_log, we use the existing Prisma model and let the trigger populate
 * prev_hash / row_hash / hash_at.
 *
 * Phase 4: swap with Valkey-stream batched writer (no API change).
 */

import type { PrismaClient } from '@prisma/client';
import {
  AuditLogInputSchema,
  CallWindowAuditInputSchema,
  ConsentLogInputSchema,
  DncSyncLogInputSchema,
  OriginateAuditInputSchema,
  type AuditLogInput,
  type CallWindowAuditInput,
  type ConsentLogInput,
  type DncSyncLogInput,
  type OriginateAuditInput,
} from './events.js';

export type AuditTable =
  | 'audit_log'
  | 'call_window_audit'
  | 'originate_audit'
  | 'consent_log'
  | 'dnc_sync_log';

export interface AppendResult {
  id: bigint;
  rowHash: string;
}

export class AuditWriter {
  constructor(private readonly db: PrismaClient) {}

  async appendAuditLog(entry: AuditLogInput): Promise<AppendResult> {
    const data = AuditLogInputSchema.parse(entry);
    // Use raw INSERT so the BEFORE INSERT trigger fires and populates row_hash.
    // Prisma's createMany/create would also work (trigger fires either way),
    // but we need LAST_INSERT_ID() + SELECT row_hash back.
    const rows = await this.db.$queryRaw<{ id: bigint; row_hash: string }[]>`
      INSERT INTO audit_log
        (tenant_id, actor_user_id, actor_kind, action, entity_type, entity_id,
         before_json, after_json, request_id, ip_address, user_agent, ts)
      VALUES (
        ${data.tenantId}, ${data.actorUserId ?? null}, ${data.actorKind},
        ${data.action}, ${data.entityType}, ${data.entityId ?? null},
        ${data.beforeJson != null ? JSON.stringify(data.beforeJson) : null},
        ${data.afterJson != null ? JSON.stringify(data.afterJson) : null},
        ${data.requestId ?? null}, ${data.ipAddress ?? null}, ${data.userAgent ?? null},
        ${data.ts}
      );
      SELECT LAST_INSERT_ID() AS id,
             (SELECT row_hash FROM audit_log WHERE id = LAST_INSERT_ID() LIMIT 1) AS row_hash
    `;
    const result = rows[0] ?? rows[rows.length - 1];
    return { id: result.id, rowHash: result.row_hash };
  }

  async appendCallWindowAudit(entry: CallWindowAuditInput): Promise<AppendResult> {
    const data = CallWindowAuditInputSchema.parse(entry);
    const rows = await this.db.$queryRaw<{ id: bigint; row_hash: string }[]>`
      INSERT INTO call_window_audit
        (tenant_id, lead_id, phone_e164, campaign_id, decision, reason,
         tz_iana, tz_confidence, state_code, zip, party_local, party_dow,
         effective_open_min, effective_close_min, rule_applied,
         enforcement_point, next_open_at, call_uuid)
      VALUES (
        ${data.tenantId}, ${data.leadId}, ${data.phoneE164}, ${data.campaignId},
        ${data.decision}, ${data.reason},
        ${data.tzIana ?? null}, ${data.tzConfidence ?? null},
        ${data.stateCode ?? null}, ${data.zip ?? null},
        ${data.partyLocal ?? null}, ${data.partyDow ?? null},
        ${data.effectiveOpenMin ?? null}, ${data.effectiveCloseMin ?? null},
        ${data.ruleApplied ?? null}, ${data.enforcementPoint},
        ${data.nextOpenAt ?? null}, ${data.callUuid ?? null}
      );
      SELECT LAST_INSERT_ID() AS id,
             (SELECT row_hash FROM call_window_audit WHERE id = LAST_INSERT_ID() LIMIT 1) AS row_hash
    `;
    const result = rows[0] ?? rows[rows.length - 1];
    return { id: result.id, rowHash: result.row_hash };
  }

  async appendOriginateAudit(entry: OriginateAuditInput): Promise<AppendResult> {
    const data = OriginateAuditInputSchema.parse(entry);
    const rows = await this.db.$queryRaw<{ id: bigint; row_hash: string }[]>`
      INSERT INTO originate_audit
        (tenant_id, attempt_uuid, call_uuid, lead_id, campaign_id, list_id, agent_id,
         mode, dial_target, carrier_id, gateway_id, gateway_name,
         caller_id_number, caller_id_source, phone_e164, originated_at,
         tcpa_decision, tcpa_reason, tcpa_tz_resolved,
         dnc_decision, dnc_sources, consent_decision, consent_state,
         bypass_token, outcome, outcome_at, duration_ms,
         error_message, fs_host, request_id, ip_address)
      VALUES (
        ${data.tenantId}, ${data.attemptUuid}, ${data.callUuid ?? null},
        ${data.leadId}, ${data.campaignId ?? null}, ${data.listId ?? null},
        ${data.agentId ?? null}, ${data.mode}, ${data.dialTarget},
        ${data.carrierId ?? null}, ${data.gatewayId ?? null},
        ${data.gatewayName ?? null}, ${data.callerIdNumber ?? null},
        ${data.callerIdSource ?? null}, ${data.phoneE164}, ${data.originatedAt},
        ${data.tcpaDecision ?? null}, ${data.tcpaReason ?? null},
        ${data.tcpaTzResolved ?? null}, ${data.dncDecision ?? null},
        ${data.dncSources != null ? JSON.stringify(data.dncSources) : null},
        ${data.consentDecision ?? null}, ${data.consentState ?? null},
        ${data.bypassToken ?? null}, ${data.outcome},
        ${data.outcomeAt ?? null}, ${data.durationMs ?? null},
        ${data.errorMessage ?? null}, ${data.fsHost ?? null},
        ${data.requestId ?? null}, ${data.ipAddress ?? null}
      );
      SELECT LAST_INSERT_ID() AS id,
             (SELECT row_hash FROM originate_audit WHERE id = LAST_INSERT_ID() LIMIT 1) AS row_hash
    `;
    const result = rows[0] ?? rows[rows.length - 1];
    return { id: result.id, rowHash: result.row_hash };
  }

  async appendConsentLog(entry: ConsentLogInput): Promise<AppendResult> {
    const data = ConsentLogInputSchema.parse(entry);
    const rows = await this.db.$queryRaw<{ id: bigint; row_hash: string }[]>`
      INSERT INTO consent_log
        (tenant_id, call_uuid, lead_id, phone_e164, prompt_id,
         dtmf_response, outcome, language, prompt_played_at)
      VALUES (
        ${data.tenantId}, ${data.callUuid}, ${data.leadId}, ${data.phoneE164},
        ${data.promptId}, ${data.dtmfResponse ?? null}, ${data.outcome},
        ${data.language}, ${data.promptPlayedAt}
      );
      SELECT LAST_INSERT_ID() AS id,
             (SELECT row_hash FROM consent_log WHERE id = LAST_INSERT_ID() LIMIT 1) AS row_hash
    `;
    const result = rows[0] ?? rows[rows.length - 1];
    return { id: result.id, rowHash: result.row_hash };
  }

  async appendDncSyncLog(entry: DncSyncLogInput): Promise<AppendResult> {
    const data = DncSyncLogInputSchema.parse(entry);
    const rows = await this.db.$queryRaw<{ id: bigint; row_hash: string }[]>`
      INSERT INTO dnc_sync_log
        (source, kind, outcome, added, removed, error_count,
         file_hash, started_at, completed_at, duration_ms, notes)
      VALUES (
        ${data.source}, ${data.kind}, ${data.outcome},
        ${data.added}, ${data.removed}, ${data.errorCount},
        ${data.fileHash ?? null}, ${data.startedAt},
        ${data.completedAt ?? null}, ${data.durationMs ?? null},
        ${data.notes ?? null}
      );
      SELECT LAST_INSERT_ID() AS id,
             (SELECT row_hash FROM dnc_sync_log WHERE id = LAST_INSERT_ID() LIMIT 1) AS row_hash
    `;
    const result = rows[0] ?? rows[rows.length - 1];
    return { id: result.id, rowHash: result.row_hash };
  }
}
